### PR TITLE
random: Initialize PRNG with random number from system.

### PR DESCRIPTION
**Pull request by fweik**
Due to popular demand and against my better judgement, initialize
the PRNG with std::random_device.

Fixes nothing

Description of changes:
 - Initialize PRNG with true random number from std::random_device.
- clang-format

PR Checklist
------------
 - [ ] Tests? Very hard to test, details are implementation specific.

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
I find this a quite dangerous interface change. And I find the current behaviour the correct one.

In my opinion any randomness in the results of a simulation script should be explicitly introduced in the script.

If there is a majority for merging this, there should be a very prominent warning in the release notes

**Comment by mkuron** [comment]
I agree with @RudolfWeeber. @fweik, is there any specific problem this is supposed to address or did you just give in to @jonaslandsgesell's repeated requests about RNG checkpointing?

**Comment by fweik** [comment]
The powers to be thought this is a good idea...

**Comment by jonaslandsgesell** [comment]
We have a python function system . set_random_state_PRNG () which allows the user to set a fully random state of the PRNG. And we also have a property "seed" in the system class which allows the user to set a seed.

The only question is whether we want a random seed as default here. In my opinion it is a good idea to do so since a simulation result should not depend on the seed which is chosen. Therefore the seed is nothing a user should have to set.
The only time a seed is intended to be set is when trying to reproduce trajectories (e.g. for test cases). However in this case I think the developer who writes the test case is responsible for using the same seed.
I think in 99% of the runs of espresso a random seed is a sensible default value.

Looks good to me. The changes are fine to merge.

@mkuron I did I requested PRNG checkpointing! Have a look there #577 
PS: I talked to Floh and this was a wish from Christian. Maybe this wish is related to this bug: #978

**Comment by RudolfWeeber** [comment]
>The only question is whether we want a random seed as default here. In my opinion it is a good idea to do so since a simulation result should not >depend on the seed which is chosen.

If the result does not depend on the seed, then 1 is just fine as choice.

 

 

>The only time a seed is intended to be set is when trying to reproduce trajectories 

This is precicely, why it should not be randomly chosen. Because in that case, it is not possible to reproduce existing simulation results unless whoever generated them wrote down the randomly chosen seed.

**Comment by kosovan** [comment]
I fully agree with Rudolf here. With sufficient statistics, the result should not depend on the seed, then a fixed seed is just fine. If the user has reasons to use different seeds, there is an option to do so.

Reproducibility of results is an important issue. For example, if you encounter a rare bug which may depend on seed, it may become very hard to reproduce it.

**Comment by KaiSzuttor** [comment]
Regarding the reproducibility in the hopefully unusual case of debugging please have a look at the gromacs docu addressing this issue: http://www.gromacs.org/Documentation/Terminology/Reproducibility
So having the same seed does not at all mean that the trajectory is reproducable.
I think the default use case still is having to perform several runs for better statistics.

**Comment by kosovan** [comment]
Well, the gromacs docs explains that simulations may not be exactly reproducible mainly due to accumulation of rounding errors if you use different architectures or different versions of libraries. However, this only shows up after a big number of steps. And if you use the same computer, libraries, same random seed, it is 100% reproducible.

With randomly chosen random_seed the situation is different: the simulation becomes totally irreproducible from the very beginning. Should this be the preferred default behaviour? What is the gain? Saving one line in the script which sets the seed? Is that worth losing reproducible default behaviour?

I think that Kai's default use case is only the production stage. Before that, in the testing and tuning stage it happens that scripts crash unexpectedly, e.g. integrator crash with the "bond broken" message discussed repeatedly in the user mailing list. In the tuning stage it is very confusing if your simulation behaves randomly. And it takes a while to figure out that you need to explicitly set random seed to achieve reproducible behaviour.

File: src/core/random.hpp [hunk 1]
```diff
@@ -78,51 +78,55 @@ void init_random_seed(int seed);
...
  */
-inline double gaussian_random_cut(void){
+inline double gaussian_random_cut(void) {
   using namespace Random;
-  const double random_number=1.042267973*normal_distribution(generator);
-  
-  if ( fabs(random_number) > 2*1.042267973 ) {
-    if ( random_number > 0 ) {
-      return 2*1.042267973;
+  const double random_number = 1.042267973 * normal_distribution(generator);
```
Comment (hmenke) [inline src/core/random.hpp]: What is this number? (`1.042267973`)  It keeps appearing here but there is no documentation whatsoever explaining the purpose and origin of it.  It was originally introduced in https://github.com/espressomd/espresso/commit/8185eae2d29463bbd7a649617ea9d662d2f821a8 also without docs.

**Comment by KaiSzuttor** [comment]
I think this is a hack for too large random number leading to undesired behaviour.

On 1 Mar 2017, at 14:34, Henri Menke <notifications@github.com<mailto:notifications@github.com>> wrote:


@hmenke commented on this pull request.

________________________________

In src/core/random.hpp<https://github.com/espressomd/espresso/pull/1039#discussion_r103683395>:

>    using namespace Random;
-  const double random_number=1.042267973*normal_distribution(generator);
-
-  if ( fabs(random_number) > 2*1.042267973 ) {
-    if ( random_number > 0 ) {
-      return 2*1.042267973;
+  const double random_number = 1.042267973 * normal_distribution(generator);


What is this number? (1.042267973) It keeps appearing here but there is no documentation whatsoever explaining the purpose and origin of it. It was originally introduced in 8185eae<https://github.com/espressomd/espresso/commit/8185eae2d29463bbd7a649617ea9d662d2f821a8> also without docs.

**Comment by hmenke** [comment]
This should be merged already.

**Comment by RudolfWeeber** [comment]
I would like to discuss this in an Espresso meeting, please.

**Comment by fweik** [comment]
Please don't merge. We will come to a decision in the next offline Espresso meeting on Tuesday.

**Comment by davidsean** [comment]
We did not agree on the best default behaviour between automatic and manual seeding.

The option to REQUIRE a seed before ESPresSo starts using random numbers was brought up. This forces the user to take an explicit stance on what the seed is, and removes the dangers associated with (wrong) expected behaviour.

To make it easy on the user, the error message will display the python command to seed randomly.
The test scripts will all have to be modified to include the explicit seeding.

It just occurred to me that perhaps the error message could simply state that no manual seeding was performed and that ESPresSo has chosen a seed.

**Comment by jonaslandsgesell** [comment]
There is literature on the langevin thermostat: If the seed is the same trajectories converge although initial conditions may differ. Maybe we dig a bit more into this. It was a group at LANL who described it:
http://pubs.acs.org/doi/abs/10.1021/ct800573m

Please use a random seed and not always the same seed.

**Comment by kosovan** [comment]
Jonas, I guess you meant: "please use different seeds in otherwise identical production runs".
If you want to reproduce the result, you should use the same seed.

I agree with David that requiring a seed before espresso starts using random numbers removes the dangers of unexpected behaviour. The user should be aware what the seed does, and if he/she is not aware, it is not the fault of Espresso.

**Comment by jonaslandsgesell** [comment]
@Peter As the Paper outlines the choice of an identical random seed for production runs introduces a synchronization bias in simulation data that are acquired via averaging over multiple simulation runs. This is true, even if the simulations are not identical in the beginning, for example because they differ in initial conditions (like atom velocities). As the paper states this is a risk when publishing data. We are in principle discussing whether we want espresso users to be at risk of suffering synchronization bias by default or not. Maybe you can do an experiment and base your decision on that: Ask your students after a lecture about the Layapunov instability whether they think the seed in a Langevin simulation matters if the atom positions are set randomly in the beginning. I suppose a majority of the students might vote for "No".

**Comment by fweik** [comment]
Closed in favour of #1211

**Pull request closed without merging**